### PR TITLE
PN-90: Genotypic seeding results need to be combined with phenotypic seeding.

### PR DESCRIPTION
**Pull request by veronikaslc**
Fixed.

### CODE REVIEW

**Comment by veronikaslc** [comment]
@allasm @mjshepherd, please review

**Comment by mjshepherd** [comment]
+1

**Comment by buske** [comment]
@veronikaslc This should gracefully handle the case when neither phenotypes nor genes are present. Currently, it caused solr to fail loudly:

```
Caused by: org.apache.solr.search.SyntaxError: Cannot parse '()-document:xwiki\:data.P0000001 -document:xwiki\:data.MIM*': Encountered " ")" ") "" at line 1, column 1.
```

File: similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultPatientSimilarityViewFactory.java [hunk 1]
```diff
@@ -313,48 +232,12 @@ private void setDescendantsMap(VocabularyTerm root, Map<VocabularyTerm, Collecti
 
         this.logger.info("Normalizing term frequency distribution...");
         // Normalize all the term frequencies to be a proper distribution
+        double maxFreq = termFreq.get(HP_ROOT);
```
Comment (buske) [inline similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultPatientSimilarityViewFactory.java]: This crashes the Component. `termFreq` is a `Map<VocabularyTerm, Double>`, and `HP_ROOT` is a String, so `.get` returns null.

File: similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultPatientSimilarityViewFactory.java [hunk 2]
```diff
@@ -187,175 +185,100 @@ private static double limitProb(double prob)
...
             }
         }
-        if (!ignoredSymptoms.isEmpty()) {
-            this.logger.warn(String.format("Ignored %d symptoms", ignoredSymptoms.size()));
-        }
 
         this.logger.info("Normalizing term frequency distribution...");
         // Normalize all the term frequencies to be a proper distribution
+        VocabularyTerm root = hpo.getTerm(HP_ROOT);
+	VocabularyTerm root = hpo.getTerm(HP_ROOT);
```
Comment (buske) [inline similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultPatientSimilarityViewFactory.java]: double paste?

File: similarity-search/src/main/java/org/phenotips/similarity/internal/SolrSimilarPatientsFinder.java [hunk 3]
```diff
@@ -204,4 +237,24 @@ private long count(SolrQuery query)
...
+                    q.append(" OR solved_genes:"
+                        + ClientUtils.escapeQueryChars(geneName));
+                    q.append(" OR candidate_genes:"
+                        + ClientUtils.escapeQueryChars(geneName));
+                }
+            }
+        }
+
+        if (q.indexOf(" OR ") == 1) {
+            q.delete(1, 5);
```
Comment (buske) [inline similarity-search/src/main/java/org/phenotips/similarity/internal/SolrSimilarPatientsFinder.java]: Hmm, this seems unnecessarily complex and opaque. The OR operators are not necessary, nor is specify the field names multiple times. I'll send along a suggestion.

File: similarity-search/src/main/java/org/phenotips/similarity/internal/SolrSimilarPatientsFinder.java [hunk 4]
```diff
@@ -204,4 +237,24 @@ private long count(SolrQuery query)
...
         return 0;
     }
+
+    private void appendGenesToQuery(PatientData<Map<String, String>> allGenes, StringBuilder q)
+    {
+        if (allGenes != null && allGenes.size() > 0 && allGenes.isIndexed()) {
+            for (Map<String, String> gene : allGenes) {
+                String geneName = gene.get("gene");
+                String status = gene.get("status");
+                if ("solved".equals(status) || "candidate".equals(status)) {
```
Comment (buske) [inline similarity-search/src/main/java/org/phenotips/similarity/internal/SolrSimilarPatientsFinder.java]: Empty status should count as candidate as well, to ensure this works properly before 1.3M1.

**Pull request closed without merging**